### ISSUE TITLE
How to use Elmo Embeddings (Word Vectors, Sentence Vectors)

### ISSUE THREAD

**Issue by andymancodes** [tensorflow/hub#149]
Hi,

Even after trying to work with elmo and reading about it, I am not getting how to use it. It looks like for a given sentence, i have to pass the sentence through the elmo model and then I can get the elmo embeddings? But the parameters of a neural net are fixed after training. So, why not release elmo as a set of pretrained vectors like glove, why make it so hard to use?

Where could I find a clear description of how to use elmo embeddings. Does elmo have word embeddings? Does elmo only give sentence embeddings? How can I use elmo to say get word embeddings and compare their performance against Glove or compare the performance of sentence embeddings. How can I build a matrix of word embeddings as in Glove or word2vec?

I have been hearing about elmo and want to use it but so far all the nice metrics in their paper is lost on me because I can't really test it.

Tensorflow hub is awesome btw.

Cheers!

**Comment by HimaVarsha94** [comment]
Thanks for the elaborate answer above. I am trying to use the above Elmo model and train using some custom data that I have. I made the trainable flag as True, but my weights don't change. Is there a documentation that I can look up?

**Comment by arnoegw** [comment]
Like https://tfhub.dev/google/elmo/2 says, the Elmo module only supports fine-tuning the weights for aggregating embeddings from its layers. It does not support training from scratch (new vocab and all). For a similar discussion of Universal Sentence Encoder, please see https://github.com/tensorflow/hub/issues/155

**Comment by PrashantRanjan09** [comment]
@HimaVarsha94 try having a look at : https://github.com/PrashantRanjan09/Elmo-Tutorial
See if it can help.

**Comment by HimaVarsha94** [comment]
@PrashantRanjan09 Thanks. You have two tutorials, one to use pre-trained ones and another for starting from scratch. Both I have already done. I am interested in using a pre-trained one at the beginning and training on top of it. The TensorFlow trainable flag is confusing me because I do not see the weights changing with that flag set as True.

**Comment by PrashantRanjan09** [comment]
@HimaVarsha94: There is one for incremental training as well, which addresses the issue you are facing. I also faced the same issue because when i was training with trainableflag as True, i also didnt see the weights changing. Although in the Tensorflow tutorial they mentioned that the new model hosted on hub allows that.

**Comment by Nomiluks** [comment]
Hi,
How can we train our data to obtain ELMO embeddings ?

**Comment by shikhar26** [comment]
@PrashantRanjan09 hello, i was able to train my model from scratch and was able to generate my weights file. Problem statement is to calculate the cosine similarity between sentence embedding. Please suggest the method to consume my model that gives us an aggregated result from all the layers in form of embedding for a sentence. 
we used https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md#using-elmo-interactively but the output embedding we receive is a tensor containing output from each layer of the ELMO model (which are 3) and contains vectors for each word in the sentence rather then the sentence as a whole.

**Comment by MitaliSodhi** [comment]
> @PrashantRanjan09 hello, i was able to train my model from scratch and was able to generate my weights file. Problem statement is to calculate the cosine similarity between sentence embedding. Please suggest the method to consume my model that gives us an aggregated result from all the layers in form of embedding for a sentence.
> we used https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md#using-elmo-interactively but the output embedding we receive is a tensor containing output from each layer of the ELMO model (which are 3) and contains vectors for each word in the sentence rather then the sentence as a whole.

@PrashantRanjan09 I am also trying to do exactly the same as what you mentioned. Please let me know if you could come up with a solution.

**Comment by PrashantRanjan09** [comment]
@MitaliSodhi : since you have the vectors, the sentence representation can be an average sum of the vectors along an axis or a weighted sum. You just have to figure out how can you take sum of those vectors along an axis.
I am not sure if you can get the sentence level elmo embedding or if the capability is already provided by allenai.
However, take a look at [this](https://www.tensorflow.org/api_docs/python/tf/math/reduce_sum)
It might help.

**Comment by PrashantRanjan09** [comment]
Also, this is what Mark Neumann (from Allen AI has to say):

- Elmo does have word embeddings, which are built up from character convolutions. However, when Elmo is used in downstream tasks, a contextual representation of each word is used which relies on the other words in the sentence.
- Elmo does not produce sentence embeddings, rather it produces embeddings per word "conditioned" on the context. This is why you have to run them through the model.
- You cannot "build a matrix of word embeddings" for the reasons above.

Please see the tutorial for a comprehensive overview.
[See](https://github.com/allenai/allennlp/issues/1737)

**Comment by jrmlhermitte** [comment]
Hi all,

   I think there is question on how to explicitly obtain just the embeddings. I don't think I have seen this explicitly written here or in an of the links (please correct me if wrong :-) ). I would like to know myself. Here is my attempt, based on this [test](https://github.com/tensorflow/hub/blob/master/examples/text_embeddings/export_test.py#L47) that was linked earlier. Is this correct? Please do correct me if I have missed an important tutorial somewhere, as I have not been very exhaustive (my ELMO journey has started about 30 min ago).

```py3
import tensorflow_hub as hub
import tensorflow as tf


# Load an embedding from tfhub
elmo = hub.Module("https://tfhub.dev/google/elmo/1")

# Try out some tokens as part of a sentence.
tokens = tf.constant(['This', 'is', 'a', 'test'])

embeddings = elmo(tokens)


# Extract the embeddings as part of a session
with tf.Session() as session:
    session.run(tf.tables_initializer())
    session.run(tf.global_variables_initializer())
    vectors = session.run(embeddings)


# outputs:
# In [16]: vectors.shape
# Out[16]: (4, 1024)

# One vector example:
# In [17]: vectors[0]
# Out[17]:
# array([-0.5142068 , -0.46363443,  0.06954904, ..., -0.09634193,
# -0.25016016,  0.05970452], dtype=float32)
```

**Comment by deanhoperobertson** [comment]
> The default output vector size is 1024 with 'elmo'. Is it possible to downsize the output, say 300 or 100, like word2vec?

-- @jundongq did you managed to work this out?

**Comment by saadi143** [comment]
> The default output vector size is 1024 with 'elmo'. Is it possible to downsize the output, say 300 or 100, like word2vec?

@jundongq  Did you got any answer to reduce the dimensions of elmo embedding. E.g 300 or 256

**Comment by saadi143** [comment]
> > The default output vector size is 1024 with 'elmo'. Is it possible to downsize the output, say 300 or 100, like word2vec?
> 
> -- @jundongq did you managed to work this out?

@deanhoperobertson Please guide me if you know how to reduce the emlo embedding dimensions to 300.

**Comment by Arjunsankarlal** [comment]
Hi @saadi143 @deanhoperobertson @jundongq , As far as I know, these are pretrained models released by allennlp, which is trained to output those specific dimensions. So if you really want to reduce the dimensions, you can try dimensionality reduction techniques like PCA, tSNE, UMAP. Hope this helps.

**Comment by saadi143** [comment]
@Arjunsankarlal Dimension reduction such as PSA will loose the information. Is there any way to get around 300 dimensions?

**Comment by Arjunsankarlal** [comment]
Hi @saadi143, I think the only option would be training the whole ELMo model on you own from the scratch, which I am not sure how much possible it is, based on the dataset availability and the hardware requirement! If you could tell what you are really trying to achieve/solve I could share my thoughts :)

**Comment by saadi143** [comment]
@Arjunsankarlal  I have to get embedding for my dataset tokens which is requirment of my thesis work. I am willing to use Elmo embedding. I get the 256 weights hdf5 file from Allen NLP but I am not sure how to use that file.  
And I am also want to train the whole model. I get the link from allennlp but i didn't understand how to train the model on my own dataset. If you know how can I train the model on my dataset or any useful source please guide me.

**Comment by Arjunsankarlal** [comment]
@saadi143 Sorry, I haven't trained a ELMo model from the scratch! You can start from [here](https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py). 

You can try posting a question in Allennlp repo, where you might get the right guidance for training ELMo from scratch.

**Comment by saadi143** [comment]
@Arjunsankarlal  Did you use pre-trained weights model? which is in hdf5 file?

**Comment by jeromeshan** [comment]
Hi, i have stupid question, how can i get just one vector from sentence using elmo

**Comment by deanhoperobertson** [comment]
> > > The default output vector size is 1024 with 'elmo'. Is it possible to downsize the output, say 300 or 100, like word2vec?
> > 
> > 
> > -- @jundongq did you managed to work this out?
> 
> @deanhoperobertson Please guide me if you know how to reduce the emlo embedding dimensions to 300.

Sorry @saadi143 I didnt find a solution to this.

**Comment by eduamf** [comment]
You can do it changing the config file (.json) or, if you`re using bilm-tf,  changing train_elmo.py.

The content of my config file is like this:
```
"bidirectional": true,
 "char_cnn": {"activation": "relu",
              "embedding": {"dim": 16},
              "filters": [[1, 16], [2, 32], [3, 64], [4, 128], [5, 256], [6, 512], [7, 1024]],
              "max_characters_per_token": 50,
              "n_characters": 262, 
              "n_highway": 1},
 "dropout": 0.1,
 "lstm": {"cell_clip": 3,
          "dim": 1024,
          "n_layers": 2,
          "proj_clip": 3,
          "projection_dim": 128,
          "use_skip_connections": true},
 "all_clip_norm_val": 10.0, 
 "n_epochs": 10,
 "n_train_tokens": 21100,
 "batch_size": 4,
 "n_tokens_vocab": 1136,
 "unroll_steps": 20,
 "n_negative_samples_batch": 16} 
```

**Comment by njfm0001** [comment]
As ELMo is trained on a corpus, is there any way to perform a lookup of the ten most similar words according to a given word?
